lib/ZcapClient.js

```diff
@@ -0,0 +1,262 @@
+   *   `didDocument` and `keyPairs` must be provided.
+   *
+   * @returns {ZcapClient} - The new ZcapClient instance.
+   */
+  constructor({
+    baseUrl, defaultHeaders = {}, agent, didDocument, keyPairs,
+    invocationSigner, delegationSigner
+  } = {}) {
+    this.baseUrl = baseUrl;
+    this.agent = agent;
+    this.defaultHeaders = {...DEFAULT_HEADERS, ...defaultHeaders};
+
+    // set the appropriate invocation and delegation signers
+    if(didDocument && keyPairs) {
+      const signers = getCapabilitySigners({didDocument, keyPairs});
+      this.invocationSigner = signers.invocationSigner;
+      this.delegationSigner = signers.delegationSigner;
+    } else if(invocationSigner && delegationSigner) {
+      this.invocationSigner = invocationSigner;
+      this.delegationSigner = delegationSigner;
+    } else {
+      throw new Error(
+        'Either `didDocument` and `keyPairs`, or `invocationSigner` and ' +
+        '`delegationSigner` must be provided.');
+    }
+  }
+
+  /**
+   * Delegates an Authorization Capability to a target delegate.
+   *
+   * @param {object} options - The options to use.
+   * @param {string} [options.url] - The relative URL to invoke the
+   *   Authorization Capability against, aka the `invocationTarget`. Either
+   *  `url` or `capability` must be specified.
+   * @param {string} [options.capability] - The parent capability to delegate.
+   *   Either `url` or `capability` must be specified.
+   * @param {string} options.targetDelegate - The URL identifying the entity to
+   *   delegate to.
+   * @param {string} [options.invocationTarget] - Optional invocation target
+   *   to use when narrowing a `capability`'s existing invocationTarget.
+   *   Default is to use `url` if `capability` is not provided, or
+   *   `capability.invocationTarget` if `capability` is provided.
+   * @param {string} [options.expires] - Optional expiration value for the
+   *   delegation. Default is 5 minutes after `Date.now()`.
+   * @param {string|Array} [options.allowedActions] - Optional list of allowed
+   *   actions or string specifying allowed delegated action. Default: [] -
+   *   delegate all actions.
+   *
+   * @returns {Promise<object>} - A promise that resolves to a delegated
+   *   capability.
+   */
+  async delegate({
+    url, capability, targetDelegate, invocationTarget, expires,
+    allowedActions = []
+  } = {}) {
+    let delegatedCapability;
+
+    // convert string value for allowedActions to array
+    allowedActions = (typeof allowedActions === 'string') ?
+      allowedActions = [allowedActions] : allowedActions;
+
+    // default expiration is 5 minutes in the future
+    const defaultExpires =
+      new Date(Date.now() + 5 * 60 * 1000).toISOString().slice(0, -5) + 'Z';
+
+    if(url) {
+      // generate the root capability
+      delegatedCapability = {
+        '@context': SECURITY_CONTEXT_V2_URL,
+        id: `urn:zcap:${await generateId()}`,
+        parentCapability: `urn:zcap:${url}`,
```

Comment:
Helper function? Also, we need to decide whether we are doing `urn:zcap:root:` to distinguish between a root zcap and a delegated one -- and if we are going to do `encodeURIComponent()` on the URL that is appended to the prefix.

Reply:
Done in ea90dba95576ecb4ce6ff1b81ed90d1844b39f06, including encodeURIComponent().